File: Data Structure/Array Or Vector/Find the Missing Number/SolutionByEkta.java
/*
   Program to find missing number in a given range using XOR of all the numbers
   a^0 = a
   a^a = 0
*/

import java.util.Scanner;

public class findMissingNumber {
	
	//function to find missing number using XOR
	public static int missingNum(int arr[]) {
		int XOR = 0;
		for(int i=0; i<arr.length; i++) {
			XOR ^= (arr[i] ^ (i+1));
		}
		return (XOR^(arr.length+1));
	}

	public static void main(String[] args) {
		Scanner s = new Scanner(System.in);
		System.out.println("enter N");
		int N = s.nextInt();
		
		System.out.println("enter the elements");
		int arr[] = new int[N-1];
		for(int n=0; n<N-1; n++) {
			arr[n] = s.nextInt();
		}
	
		System.out.println(missingNum(arr));
	}

}

File: Data Structure/Array Or Vector/Find The Smallest Missing Number/SolutionByEkta.cpp
/*
  program to find the smallest missing number
  time complexity O(n)
*/
#include <bits/stdc++.h>
using namespace std;

int missingNum(int arr[], int N, int M) {
    int missingNumber;
    for(int i=0; i<N && i<M; i++) {
        if(arr[i] != i){
            missingNumber = i;
            break;
        }
    }
    return missingNumber;
}

int main()
{
    int N, M, missingNumber;
    cout<<"enter size of array"<<endl;
    cin>>N;
    int arr[N];
    cout<<"enter elements of array"<<endl;
    for(int n=0; n<N; n++){
        cin>>arr[n];
    }
    cout<<"enter M"<<endl;
    cin>>M;

    if(arr[N-1] == N-1){
        missingNumber = N;
    }else{
        missingNumber = missingNum(arr, N, M);
    }
    cout <<"smallest missing number is : "<<missingNumber<< endl;
    return 0;
}
